plugins/touchegg/Main.vala

```diff
@@ -0,0 +1,159 @@
+            return false;
+        }));
+        client.on_gesture_end.connect ((gesture) => Idle.add (() => {
+            on_handle_gesture (gesture, "end");
+            return false;
+        }));
+
+        try {
+            client.run ();
+        } catch (Error e) {
+            warning ("Error initializing Touchégg client: %s", e.message);
+        }
+    }
+
+    public override void destroy () {
+        if (client != null) {
+            client.stop ();
+        }
+    }
+
+
+    private void on_handle_gesture (Gesture gesture, string event) {
+        // debug (@"Gesture $(event): $(gesture.type) - $(gesture.direction) - $(gesture.fingers) fingers - $(gesture.percentage)% - $(gesture.elapsed_time) - $(gesture.performed_on_device_type)");
+        var hints = build_hints_from_gesture (gesture, event);
+
+        if (is_open_workspace_gesture (gesture)) {
+            wm.workspace_view.open (hints);
+        }
+
+        if (is_close_workspace_gesture (gesture)) {
+            wm.workspace_view.close (hints);
+        }
+
+        if (is_next_desktop_gesture (gesture)) {
+            if (!wm.workspace_view.is_opened ()) {
+                wm.switch_to_next_workspace (Meta.MotionDirection.RIGHT, hints);
+            }
+        }
+
+        if (is_previous_desktop_gesture (gesture)) {
+            if (!wm.workspace_view.is_opened ()) {
+                wm.switch_to_next_workspace (Meta.MotionDirection.LEFT, hints);
+            }
+        }
+    }
+
+    private GLib.HashTable<string, Variant> build_hints_from_gesture (Gesture gesture, string event) {
+        var hints = new GLib.HashTable<string, Variant> (str_hash, str_equal);
+        hints.insert ("manual_animation", new Variant.boolean (true));
+        hints.insert ("event", new Variant.string (event));
+        hints.insert ("percentage", new Variant.int32 (gesture.percentage));
+
+        if (event == "end") {
+            hints.insert ("cancel_action", new Variant.boolean (gesture.percentage < SUCCEESS_THRESHOLD));
+        }
+
+        return hints;
+    }
+
+    private bool is_open_workspace_gesture (Gesture gesture) {
+        bool enabled = gala_settings.get_boolean ("multitasking-gesture-enabled");
+        int fingers = gala_settings.get_int ("multitasking-gesture-fingers");
+
+        return enabled
+            && gesture.type == GestureType.SWIPE
+            && gesture.direction == GestureDirection.UP
+            && gesture.fingers == fingers;
+    }
+
+    private bool is_close_workspace_gesture (Gesture gesture) {
+        bool enabled = gala_settings.get_boolean ("multitasking-gesture-enabled");
+        int fingers = gala_settings.get_int ("multitasking-gesture-fingers");
+
+        return enabled
+            && gesture.type == GestureType.SWIPE
+            && gesture.direction == GestureDirection.DOWN
+            && gesture.fingers == fingers;
+    }
+
+    private bool is_next_desktop_gesture (Gesture gesture) {
```

Comment:
I feel like `is_next_desktop_gesture` and `is_previous_desktop_gesture` could be combined into one method that returns a bool and the direction by an `out` parameter like this:
```
private bool is_switch_desktop_gesture (Gesture gesture, out Meta.MotionDirection dir) {
    ...

    if (enabled && gesture.type == GestureType.SWIPE && gesture.fingers == fingers) {
         if (gesture.direction == GestureDirection.LEFT) {
             dir = natural_scroll ? Meta.MotionDirection.RIGHT : Meta.MotionDirection.LEFT;
             return true;
         } else if (gesture.direction == GestureDirection.RIGHT) {
             dir = natural_scroll ? Meta.MotionDirection.LEFT : Meta.MotionDirection.RIGHT;
             return true;
         }
    }

    return false;
}
```

And then use:
```
Meta.MotionDirection dir;
if (is_switch_desktop_gesture (gesture, out dir)) {
    if (!wm.workspace_view.is_opened ()) {
        wm.switch_to_next_workspace (dir, hints);
    }
}
````

Reply:
Because the methods are rather simple and this only avoids duplicating a couple of bool checks, I'll rather leave as it is. In my opinion is more readable and avoids returning 2 arguments from a function. But feel free to modify it if you really don't like it.

Reply:
Absolutely, this was just a suggestion and nothing that I feel strongly about. The separate versions are easier to reason about for sure.